File: services/title-normalizer/title_normalizer/pipeline.py
"""Turns a job payload (raw provider items) into master media records with variants. Pure: no I/O."""

import hashlib
import re
from collections import Counter
from dataclasses import dataclass, replace
from typing import Any

from title_normalizer import tags
from title_normalizer.matching import group_titles
from title_normalizer.parser import ParsedTitle, parse_title, parse_year


@dataclass(frozen=True)
class Variant:
    stream_id: str
    raw_title: str
    label: str
    quality: str | None
    source: str | None
    audio_languages: tuple[str, ...]
    audio_tag: str | None
    is_hdr: bool
    quality_score: int
    category_id: str | None
    poster_url: str | None
    rating: float | None
    container_extension: str | None
    added_at: int | None
    release_key: int | None
    subtitle_languages: tuple[str, ...] = ()


@dataclass(frozen=True)
class Master:
    id: str
    title: str
    normalized_key: str
    year: int | None
    poster_url: str | None
    rating: float | None
    best_quality: str | None
    added_at: int | None
    release_key: int | None
    variants: tuple[Variant, ...]


def build_masters(account_id: str, media_kind: str, items: list[dict[str, Any]]) -> list[Master]:
    """`items` use the backend payload shape: id, name, categoryId, posterUrl, rating, containerExtension, releaseDate, addedAt,
    tmdbId."""
    usable = [item for item in items if str(item.get("id") or "").strip() and str(item.get("name") or "").strip()]
    parsed = [_parse_item(item) for item in usable]
    groups = merge_by_tmdb(group_titles(parsed), [tmdb_id(item) for item in usable], [title.year for title in parsed])
    masters = [_build_master(account_id, media_kind, [usable[i] for i in group], [parsed[i] for i in group]) for group in groups]
    return sorted(masters, key=lambda master: (master.title.lower(), master.year or 0))


def tmdb_id(item: dict[str, Any]) -> str | None:
    """The TMDB id some providers send in their lists (`tmdb` or `tmdb_id`); "0" and empty mean none."""
    value = str(item.get("tmdbId") or "").strip()
    return value if value.isdigit() and value.strip("0") else None


def merge_by_tmdb(groups: list[list[int]], tmdb_ids: list[str | None], years: list[int | None]) -> list[list[int]]:
    """Joins name groups that share a TMDB id, e.g. "La Casa de Papel" and "Money Heist" (D-065). A shared id only joins
    groups whose years agree (or are unknown): providers sometimes reuse an id for a remake or get it wrong."""
    parent = list(range(len(groups)))

    def find(index: int) -> int:
        while parent[index] != index:
            parent[index] = parent[parent[index]]
            index = parent[index]
        return index

    group_years = [{years[i] for i in group if years[i] is not None} for group in groups]
    first_group: dict[str, int] = {}
    for index, group in enumerate(groups):
        for identifier in sorted({tmdb_ids[i] for i in group if tmdb_ids[i]}):
            other = first_group.setdefault(identifier, index)
            if other == index:
                continue
            root, other_root = find(index), find(other)
            known = group_years[root], group_years[other_root]
            if root != other_root and (not known[0] or not known[1] or known[0] & known[1]):
                parent[max(root, other_root)] = min(root, other_root)
                group_years[min(root, other_root)] = known[0] | known[1]

    merged: dict[int, list[int]] = {}
    for index, group in enumerate(groups):
        merged.setdefault(find(index), []).extend(group)
    return [sorted(indexes) for indexes in merged.values()]


def quality_score(title: ParsedTitle) -> int:
    score = tags.QUALITY_RANK.get(title.quality or "", tags.UNKNOWN_QUALITY_RANK)
    score += tags.SOURCE_ADJUSTMENT.get(title.source or "", 0)
    score += 5 if title.is_hdr else 0
    return max(score, 0)


def variant_label(title: ParsedTitle, container_extension: str | None) -> str:
    parts = [
        title.quality,
        title.source if title.source in ("CAM", "TS", "TC", "SCR", "REMUX") else None,
        "HDR" if title.is_hdr else None,
        "/".join(title.audio_languages) or None,
        title.audio_tag,
    ]
    label = " · ".join(part for part in parts if part)
    return label or (container_extension or "Standard").upper()


def release_key(release_date: str | None, year: int | None) -> int | None:
    """YYYYMMDD from an ISO-like date ("2020-05-12"), else YYYY0000 from the year, for sorting."""
    match = re.match(r"\s*(\d{4})-(\d{1,2})-(\d{1,2})", release_date or "")
    if match and parse_year(match[1]) and 1 <= int(match[2]) <= 12 and 1 <= int(match[3]) <= 31:
        return int(match[1]) * 10000 + int(match[2]) * 100 + int(match[3])
    return year * 10000 if year else None


def master_id(account_id: str, media_kind: str, compact_key: str, year: int | None) -> str:
    """Stable across re-syncs while the group's key and year stay the same."""
    digest = hashlib.sha1(f"{account_id}|{media_kind}|{compact_key}|{year or ''}".encode()).hexdigest()
    return digest[:20]


def _parse_item(item: dict[str, Any]) -> ParsedTitle:
    parsed = parse_title(str(item["name"]))
    if parsed.year is None and (release_year := parse_year(str(item.get("releaseDate") or "")[:4])):
        parsed = replace(parsed, year=release_year)
    return parsed


def _build_master(account_id: str, media_kind: str, items: list[dict[str, Any]], parsed: list[ParsedTitle]) -> Master:
    variants = [_build_variant(item, title) for item, title in zip(items, parsed, strict=True)]
    order = sorted(range(len(variants)), key=lambda i: (-variants[i].quality_score, variants[i].stream_id))
    variants = _dedupe_labels([variants[i] for i in order])
    parsed = [parsed[i] for i in order]

    title_counts = Counter(title.clean_title for title in parsed)
    display_title = max(title_counts, key=lambda name: (title_counts[name], -[t.clean_title for t in parsed].index(name)))
    years = Counter(title.year for title in parsed if title.year is not None)
    year = years.most_common(1)[0][0] if years else None
    canonical = next(title for title in parsed if title.clean_title == display_title)
    ratings = [variant.rating for variant in variants if variant.rating is not None]
    qualities = [variant.quality for variant in variants if variant.quality]
    added = [variant.added_at for variant in variants if variant.added_at is not None]
    released = [variant.release_key for variant in variants if variant.release_key is not None]

    return Master(
        id=master_id(account_id, media_kind, canonical.compact_key, year),
        title=display_title,
        normalized_key=canonical.key,
        year=year,
        poster_url=next((variant.poster_url for variant in variants if variant.poster_url), None),
        rating=max(ratings) if ratings else None,
        best_quality=max(qualities, key=lambda q: tags.QUALITY_RANK.get(q, 0)) if qualities else None,
        added_at=max(added) if added else None,
        release_key=min(released) if released else release_key(None, year),
        variants=tuple(variants),
    )


def _build_variant(item: dict[str, Any], title: ParsedTitle) -> Variant:
    container = _optional_str(item.get("containerExtension"))
    rating = item.get("rating")
    added_at = item.get("addedAt")
    return Variant(
        stream_id=str(item["id"]).strip(),
        raw_title=str(item["name"]),
        label=variant_label(title, container),
        quality=title.quality,
        source=title.source,
        audio_languages=title.audio_languages,
        audio_tag=title.audio_tag,
        is_hdr=title.is_hdr,
        quality_score=quality_score(title),
        category_id=_optional_str(item.get("categoryId")),
        poster_url=_optional_str(item.get("posterUrl")),
        rating=float(rating) if isinstance(rating, (int, float)) else None,
        container_extension=container,
        added_at=int(added_at) if isinstance(added_at, (int, float)) and added_at > 0 else None,
        release_key=release_key(_optional_str(item.get("releaseDate")), None),
        subtitle_languages=title.subtitle_languages,
    )


def _dedupe_labels(variants: list[Variant]) -> list[Variant]:
    """Identical labels get " (2)", " (3)" so the version selector stays unambiguous."""
    seen: Counter[str] = Counter()
    result = []
    for variant in variants:
        seen[variant.label] += 1
        count = seen[variant.label]
        result.append(variant if count == 1 else replace(variant, label=f"{variant.label} ({count})"))
    return result


def _optional_str(value: Any) -> str | None:
    text = str(value).strip() if value is not None else ""
    return text or None

File: packages/shared/src/direct/normalizer/pipeline.ts
import { groupTitles, groupTitlesAsync } from './matching';
import { compactKey, parseTitle, parseYear, type ParsedTitle } from './parser';
import { sha1Hex } from './sha1';
import * as tags from './tags';

/** Raw provider items → master titles with variants. Port of title_normalizer/pipeline.py (D-038). Pure: no I/O. */
export interface NormalizerItem {
  id?: unknown;
  name?: unknown;
  categoryId?: string | null;
  posterUrl?: string | null;
  rating?: number | null;
  containerExtension?: string | null;
  releaseDate?: string | null;
  /** Unix seconds the provider added (movies) or last changed (series) the item. */
  addedAt?: number | null;
}

export interface Variant {
  streamId: string;
  rawTitle: string;
  label: string;
  quality: string | null;
  source: string | null;
  audioLanguages: string[];
  audioTag: string | null;
  isHdr: boolean;
  qualityScore: number;
  categoryId: string | null;
  posterUrl: string | null;
  rating: number | null;
  containerExtension: string | null;
  /** From the name ("SUB ITA", "VOSTFR"); `MULTI` = several (D-063). */
  subtitleLanguages: string[];
}

export interface Master {
  id: string;
  title: string;
  normalizedKey: string;
  year: number | null;
  posterUrl: string | null;
  rating: number | null;
  bestQuality: string | null;
  /** Newest `addedAt` among the variants (Unix seconds). */
  addedAt: number | null;
  /** Release date as YYYYMMDD for sorting; a year-only date is YYYY0000. */
  releaseKey: number | null;
  variants: Variant[];
}

const text = (value: unknown) => (value === null || value === undefined ? '' : String(value).trim());
const optional = (value: unknown) => text(value) || null;
const rank = (quality: string) => tags.QUALITY_RANK[quality] ?? 0;
/** Python compares str by code point; localeCompare would not. */
const compare = (a: string, b: string) => (a < b ? -1 : a > b ? 1 : 0);

export function buildMasters(accountId: string, mediaKind: string, items: NormalizerItem[]): Master[] {
  const usable = items.filter((item) => text(item.id) && text(item.name));
  return assemble(accountId, mediaKind, usable, usable.map(parseItem));
}

/**
 * Same result as `buildMasters`, but works in chunks and yields between them so the UI stays responsive, and
 * `onProgress(done, total)` reports how far it got (direct mode on TV/phone, D-038). `done` runs from 0 to `total`
 * across all steps: reading the names (first half), matching them (to 80 %), building the titles (the rest).
 */
export async function buildMastersInChunks(
  accountId: string,
  mediaKind: string,
  items: NormalizerItem[],
  { chunkSize = 500, onProgress }: { chunkSize?: number; onProgress?(done: number, total: number): void } = {},
): Promise<Master[]> {
  const usable = items.filter((item) => text(item.id) && text(item.name));
  const total = usable.length;
  const pause = (fraction: number) => {
    onProgress?.(Math.min(total, Math.floor(fraction * total)), total);
    return new Promise<void>((resolve) => setTimeout(resolve, 0));
  };

  const parsed: ParsedTitle[] = [];
  for (let start = 0; start < total; start += chunkSize) {
    for (const item of usable.slice(start, start + chunkSize)) parsed.push(parseItem(item));
    await pause((0.5 * parsed.length) / total);
  }
  const groups = await groupTitlesAsync(parsed, (done) => pause(0.5 + 0.3 * done));
  const masters: Master[] = [];
  for (let start = 0; start < groups.length; start += chunkSize) {
    for (const group of groups.slice(start, start + chunkSize)) masters.push(masterOf(accountId, mediaKind, usable, parsed, group));
    await pause(0.8 + (0.2 * Math.min(groups.length, start + chunkSize)) / groups.length);
  }
  onProgress?.(total, total);
  return sortMasters(masters);
}

function assemble(accountId: string, mediaKind: string, usable: NormalizerItem[], parsed: ParsedTitle[]): Master[] {
  return sortMasters(groupTitles(parsed).map((group) => masterOf(accountId, mediaKind, usable, parsed, group)));
}

const masterOf = (accountId: string, mediaKind: string, usable: NormalizerItem[], parsed: ParsedTitle[], group: number[]) =>
  buildMaster(
    accountId,
    mediaKind,
    group.map((index) => usable[index]!),
    group.map((index) => parsed[index]!),
  );

/** By title (case-insensitive), then year. Lower-cased once per title, not once per comparison. */
function sortMasters(masters: Master[]): Master[] {
  const keyed = masters.map((master) => ({ master, title: master.title.toLowerCase() }));
  keyed.sort((a, b) => compare(a.title, b.title) || (a.master.year ?? 0) - (b.master.year ?? 0));
  return keyed.map(({ master }) => master);
}

export function qualityScore(title: ParsedTitle): number {
  let score = title.quality ? (tags.QUALITY_RANK[title.quality] ?? tags.UNKNOWN_QUALITY_RANK) : tags.UNKNOWN_QUALITY_RANK;
  score += title.source ? (tags.SOURCE_ADJUSTMENT[title.source] ?? 0) : 0;
  score += title.isHdr ? 5 : 0;
  return Math.max(score, 0);
}

export function variantLabel(title: ParsedTitle, containerExtension: string | null): string {
  const parts = [
    title.quality,
    title.source && ['CAM', 'TS', 'TC', 'SCR', 'REMUX'].includes(title.source) ? title.source : null,
    title.isHdr ? 'HDR' : null,
    title.audioLanguages.join('/') || null,
    title.audioTag,
  ];
  return parts.filter(Boolean).join(' · ') || (containerExtension || 'Standard').toUpperCase();
}

/** YYYYMMDD from an ISO-like date ("2020-05-12"), else YYYY0000 from the year, for sorting. */
export function releaseKey(releaseDate: string | null, year: number | null): number | null {
  const match = /^\s*(\d{4})-(\d{1,2})-(\d{1,2})/.exec(releaseDate ?? '');
  if (match && parseYear(match[1]!) && +match[2]! >= 1 && +match[2]! <= 12 && +match[3]! >= 1 && +match[3]! <= 31)
    return +match[1]! * 10000 + +match[2]! * 100 + +match[3]!;
  return year ? year * 10000 : null;
}

/** Stable across re-syncs while the group's key and year stay the same. Same hash as the Python normalizer. */
export const masterId = (accountId: string, mediaKind: string, key: string, year: number | null) =>
  sha1Hex(`${accountId}|${mediaKind}|${key}|${year ?? ''}`).slice(0, 20);

function parseItem(item: NormalizerItem): ParsedTitle {
  const parsed = parseTitle(String(item.name));
  const releaseYear = parsed.year === null ? parseYear(text(item.releaseDate).slice(0, 4)) : null;
  return releaseYear ? { ...parsed, year: releaseYear } : parsed;
}

function buildMaster(accountId: string, mediaKind: string, items: NormalizerItem[], parsed: ParsedTitle[]): Master {
  const built = items.map((item, index) => buildVariant(item, parsed[index]!));
  const order = built
    .map((_, index) => index)
    .sort((a, b) => built[b]!.qualityScore - built[a]!.qualityScore || compare(built[a]!.streamId, built[b]!.streamId));
  const variants = dedupeLabels(order.map((index) => built[index]!));
  const ordered = order.map((index) => parsed[index]!);

  // Most common spelling wins; ties go to the spelling seen first (best variant first).
  const titleCounts = countInOrder(ordered.map((title) => title.cleanTitle));
  const displayTitle = [...titleCounts].reduce((best, entry) => (entry[1] > best[1] ? entry : best))[0];
  const yearCounts = countInOrder(ordered.flatMap((title) => (title.year === null ? [] : [title.year])));
  const year = yearCounts.size ? [...yearCounts].reduce((best, entry) => (entry[1] > best[1] ? entry : best))[0] : null;
  const canonical = ordered.find((title) => title.cleanTitle === displayTitle)!;
  const ratings = variants.flatMap((variant) => (variant.rating === null ? [] : [variant.rating]));
  const qualities = variants.flatMap((variant) => (variant.quality ? [variant.quality] : []));
  const added = items.flatMap((item) => (typeof item.addedAt === 'number' && item.addedAt > 0 ? [Math.trunc(item.addedAt)] : []));
  const released = items.flatMap((item) => releaseKey(optional(item.releaseDate), null) ?? []);

  return {
    id: masterId(accountId, mediaKind, compactKey(canonical), year),
    title: displayTitle,
    normalizedKey: canonical.key,
    year,
    posterUrl: variants.find((variant) => variant.posterUrl)?.posterUrl ?? null,
    rating: ratings.length ? Math.max(...ratings) : null,
    bestQuality: qualities.length ? qualities.reduce((best, quality) => (rank(quality) > rank(best) ? quality : best)) : null,
    addedAt: added.length ? Math.max(...added) : null,
    releaseKey: released.length ? Math.min(...released) : releaseKey(null, year),
    variants,
  };
}

function buildVariant(item: NormalizerItem, title: ParsedTitle): Variant {
  const container = optional(item.containerExtension);
  return {
    streamId: text(item.id),
    rawTitle: String(item.name),
    label: variantLabel(title, container),
    quality: title.quality,
    source: title.source,
    audioLanguages: title.audioLanguages,
    audioTag: title.audioTag,
    isHdr: title.isHdr,
    qualityScore: qualityScore(title),
    categoryId: optional(item.categoryId),
    posterUrl: optional(item.posterUrl),
    rating: typeof item.rating === 'number' ? item.rating : null,
    containerExtension: container,
    subtitleLanguages: title.subtitleLanguages,
  };
}

/** Identical labels get " (2)", " (3)" so the version selector stays unambiguous. */
function dedupeLabels(variants: Variant[]): Variant[] {
  const seen = new Map<string, number>();
  return variants.map((variant) => {
    const count = (seen.get(variant.label) ?? 0) + 1;
    seen.set(variant.label, count);
    return count === 1 ? variant : { ...variant, label: `${variant.label} (${count})` };
  });
}

function countInOrder<T>(values: T[]): Map<T, number> {
  const counts = new Map<T, number>();
  for (const value of values) counts.set(value, (counts.get(value) ?? 0) + 1);
  return counts;
}
